### PR TITLE
State synchronization fix

### PR DESCRIPTION
**Pull request by dchambers**
This pull-request has a soft-dependency on #4 and #5. Whereas those pull-requests prevent `render()` and `shallowRender()` from continually creating new state, this pull-request causes the state produced by `render()` and `shallowRender()` to be synchronized, causing it to seem like there actually is only one render state in-play.

It allows tests that use the rendered DOM to perform behaviour changes, to subsequently use a shallow render to confirm those changes (or vice versa), for example:

```jsx
it('updates the footer information when the completed filter is clicked', function() {
  // given
  let todoApp = $(<TodoApp model={model} router={router}/>);

  // when
  todoApp.render().find('.completed').trigger('click');

  // then
  expect(todoApp.shallowRender()[0], 'to contain',
    <TodoFooter nowShowing="completed"/>
  );
});
```

At present, this pull-request requires the test author to load a modified version of `React` as follows:

```es6
import $ from 'teaspoon';
import _React from 'react';
const React = $.syncReact(_React);
```

My knowledge of React is quite poor, but in theory I believe it should be possible to hide this from the end-developer, and instead have _teaspoon_ automatically re-create the instantiated React components it receives, but where it uses `synchronizedComponent()` underneath the hood, so that tests automatically create synchronized versions of those components.

I'm encouraged to believe this may be possible since I can add the following snippet of code to `$.fn.init()` within `QueryCollection` without breaking any tests:

```js
if(element instanceof React.Component) {
  console.log('>>> CLONED');
  element = React.cloneElement(element);
}
```

This code is causing a completely new React component, with identical children and props to be created. If you could additionally swap the type with a synchronized version, then these details could be completely hidden from the developer.

### CODE REVIEW

**Comment by dchambers** [comment]
> I'm encouraged to believe this may be possible since I can add the following snippet of code to `$.fn.init()` within `QueryCollection` without breaking any tests:

Scratch that. This was only working because I put my code snippet below this line of code:

```es6
let elements = element == null ? [] : [].concat(element);
```

and as soon as I move it above that line then all the tests break. It now makes far more sense to me why `React.cloneElement()` was able to return something that didn't seem particularly comparable yet the tests still worked.

**Comment by jquense** [comment]
hey there, thanks for the PR.

I appreciate the work you put in here, but I am not sure this is the correct approach...I actually find this behavior of maintaining state to be unintuitive behavior, compared to just caching the result of a `render()` call instead of the initial element. `let $inst = $(<Counter/>).render()`. besides that though
 there are a few issues with this approach

- it makes it impossible to actually rerender a component fresh (like the original behavior)
- it makes the "parent" collection stateful (holding the renderer and instance) while not being very clear
- the `$.syncReact` api is not great, and looks like it would could create a lot of smaller unrelated problems
- shallowRender and normal render are really orthagonal processes, the componets don't always have a simple one to one analog, what about state in children? what about when shallowRender doesn't render those children, but render() does?
- It's not clear _why_ you'd need to sync state between a normal and shallow render, generally thse are going to be completely unrelated things...if you already rendered the Component, there is no need to then shallow render it to check that component state, you already have the rendered component.

Ultimately I would just refactor the example to sometihng like this:

```js
let inst = $(<Counter/>).render();

inst[0].increment()
inst[0].state.count.should.equal(1)

inst.text().should.equal('1')
```

**Comment by dchambers** [comment]
Hi @jquense,

I've raised three related pull-requests. Of the three, this one is currently a bit of a pig in terms of the API changes it introduces. Before I attempt to address the points you've made, it might be worth first confirming whether you're happy with #4 and #5 first?

**Maintain state between renders (shallow render)**
#4 is the least controversial of all. Without it it's impossible to shallow render a component with anything but the component's initial state. Are you happy that this one makes sense?

**Maintain state between renders (full render)**
#5 isn't strictly necessary, and is really more of convenience. In tests where I need to refer to the rendered output multiple times, this change allows me to do that without memoizing the output of `render()` first, so it makes it possible to write more BDD style spec tests, as opposed to ` style unit tests. For example, I can write this:

```jsx
// when
game.render().find('.torpedo1').trigger('click');
game.render().find('.torpedo2').trigger('click');
```

instead of this:

```jsx
// when
let gameRender = game.render();
gameRender.find('.torpedo1').trigger('click');
gameRender.find('.torpedo2').trigger('click');
```

This change does mean that `render()` won't do a full re-render without also using `$()` again, which is somewhat controversial, but I'm not personally aware of any use-case for doing a re-render given that the rendered virtual DOM is _live_ anyway, and is automatically updated as the children, props or state change.

Perhaps you have some use cases for needing this you can share?

**State Synchronization Fix**

That finally brings us to this pull-request and its ugly API. From your feedback, I think the most important thing to share with you first is a valid use case for this change. I've created a [react-todomvc](https://github.com/dchambers/react-todomvc) repo containing just the React version of the TodoMVC app, but where I've added [some tests](https://github.com/dchambers/react-todomvc/blob/master/test/TodoAppTest.jsx), and where the app has been modified to make it more amenable to testing.

If you look at this [commit](https://github.com/dchambers/react-todomvc/commit/f8efd6e154fc7c3056fba29d9042c26ace3e07d4) you will see that some of the tests (those that involve state) previously had to be `xit`ed because they aren't possible with teaspoon as it currently stands. This is because I write behaviour tests using this form:

```
Given: Input Form
When: Behaviour
Then: Output Form
```

Here's an example test:

```jsx
it('displays a Clear Completed button when at least one item is marked as done', function() {
  // given
  var todoApp = $(<TodoApp model={model} router={router}/>);

  // when
  var route = extractRoute(todoApp.render().find('a.completed').dom());
  router.dispatch('on', route);

  // then
  expect(todoApp.shallowRender()[0], 'to contain',
    <TodoFooter count={1} completedCount={0} nowShowing="completed"/>
  );
```

Here, the _when_ is using the rendered form, whereas the _then_ is using the shallow rendered form.

The shallow form is used for the _then_ because it allows components to define their output using semantically rich tags, which makes tests easier to comprehend, and less brittle, since changes to the DOM for a sub-component (which is private implementation detail) will not cause the test to break. The full rendered form is used for behaviours because it creates a far more authentic test, and means we have component tests which are good enough that no QA will be tempted to disregard them as unit tests only, and then start adding unreliable Selenium tests to get the coverage they desire.

You'll notice that the _when_ and the _then_ both involve a single component created in the _given_. Logically then, there is only one component under test, and provided this component limits itself to altering only its props and children, then everything works in the way the reader would expect. However, because a component's state is maintained by the React renderer rather than the component itself, things cease to make sense as soon as the component's state changes. Although there is only component instance, there are two component states in-play (one for `render()` and one for `shallowRender()`), and so we get spurious results.

This pull-request is therefore a patch to overcome React's decision to maintain the state in the renderer, allowing the end-developer to test a single instance of the component as though the state for that instance was maintained within the component itself.

Apart from the ugliness of the API though, I don't think the behaviour is particularly controversial. If I'd been able to achieve the API I really wanted (where `$()` would fix-up the state synchronization issue for a given component instance) then this aspect would be fixed too. I might have another try at this at some point, but could otherwise move this code into a separate library since it doesn't meld well will teaspoon at present anyway.

**Your Points**

Finally, to get back to the points you made:

>   * it makes it impossible to actually rerender a component fresh (like the original behavior)

Yes, agreed. I'm not aware this is ever needed though, and I'm happy to drop #5 as it's only a convenience if you are aware of why it might be needed.

>   * it makes the "parent" collection stateful (holding the renderer and instance) while not being very clear

Yes, but this seems correct to me.

>   * the `$.syncReact` api is not great, and looks like it would could create a lot of smaller unrelated problems

Strongly agree that it's ugly! It may cause other problems too, but for me, at present, it's helping.

>   * `shallowRender` and normal `render` are really orthagonal processes, the components don't always have a simple one to one analog, what about state in children? what about when `shallowRender` doesn't render those children, but `render()` does?

If a component isn't relevant for both renderers than it didn't need to be synchronized in the first place, so there was no problem from the get go.

>   * It's not clear why you'd need to sync state between a normal and shallow render, generally thse are going to be completely unrelated things...if you already rendered the Component, there is no need to then shallow render it to check that component state, you already have the rendered component.

Hopefully I've now explained why this does actually make sense.

> Ultimately I would just refactor the example to something like this:

> ```jsx
let inst = $(<Counter/>).render();

> inst[0].increment()
inst[0].state.count.should.equal(1)

> inst.text().should.equal('1')
```

I think the `<Counter/>` example is actually too simple to be instructive. How about an example like the one I use above. The only solution here is to avoid using shallow rendering, which leads to harder to comprehend and more brittle tests.

Basically, as it stands, shallow rendering can't be used for tests that involve component state.

**Comment by dchambers** [comment]
I've decided to close this as it makes more sense to move this functionality into an external library until/if a way can be found to do state synchronization completely transparently.

**Comment by jquense** [comment]
first thanks for taking the time to lay out your thoughts.

>Although there is only component instance, there are two component states in-play (one for render() and one for shallowRender()), and so we get spurious results.

You don't actually have one instance though, you have one element, that is only a description of a potential instance. Its not an instance until you render it. That you end up with two component states is correct to my mind, analogous to how creating two windows/tabs in a browser of the same html creates two unshared contexts.

I still think that this approach is really counter to the intent of either of these mechanisms, as evidenced by the need to _override_ `createElement` in order to even get this behavior. Its not that teaspoon makes this difficult its that react does. I appreciate your example but I think you are conflating concepts in your testing. normal render doesn't require that you write integration like brittle tests, any more than shallowRender somehow protects you from it. In your example:

```js
it('displays a Clear Completed button when at least one item is marked as done', function() {
  // given
  var todoApp = $(<TodoApp model={model} router={router}/>).render();

  // when
  var route = extractRoute(todoApp.find('a.completed').dom());
  router.dispatch('on', route);

  // then
todoApp.find(TodoFooter)[0].props.should.eql({ 
  count: 1, 
  completedCount: 0, 
  nowShowing: 'completed' 
})
```

If you don't want to test DOM configurations then don't, you don't need to rerender the entire component again using a new render substrate (shallowRender) to do that though.

It seems like you like shallow rendering because its _easier_ to write declarative asserts, which I can appreciate but in that case it is probably going to be less error prone/hacky to write an api that allows for that specifically instead of requiring you rerender the entire component shallowly when the component is already rendered such as:

```js
expect(todoApp.find(TodoFooter).element(), 'to equal'',
    <TodoFooter count={1} completedCount={0} nowShowing="completed"/>
);
```

**Comment by dchambers** [comment]
Hi @jquense, thanks for coming back to me on this. Your understanding about an instantiated React component being an _element_ and not an acuated instance is particularly interesting, although I do still have very good reasons for doing what I'm doing, even if it's not easy to do :-)

Could you please just take a look at #4, since that one's an obvious bug with teaspoon as it stands &mdash; not allowing you to test stateful components using shallow rendering.

**Comment by jquense** [comment]
just to clarify a bit, jsx elements not being instances isn't actually my understanding, that's Reacts definition of them, you only get instances of components when they are rendered, u till then they are just descriptors (in fact they used to be called that up until react 0.12)

more to this issue tho does my ending API alternative proposal help your use case?

**Comment by dchambers** [comment]
> just to clarify a bit, jsx elements not being instances isn't actually my understanding, that's Reacts definition of them, you only get instances of components when they are rendered, u till then they are just descriptors (in fact they used to be called that up until react 0.12)

I don't doubt you, I just didn't know, which is why I found it interesting! In fact, I only started developing with React nine days ago, so I've pretty much everything still to learn.

> more to this issue tho does my ending API alternative proposal help your use case?

If you're talking about this:

```
it('displays a Clear Completed button when at least one item is marked as done', function() {
  // given
  var todoApp = $(<TodoApp model={model} router={router}/>).render();

  // when
  var route = extractRoute(todoApp.find('a.completed').dom());
  router.dispatch('on', route);

  // then
todoApp.find(TodoFooter)[0].props.should.eql({ 
  count: 1, 
  completedCount: 0, 
  nowShowing: 'completed' 
})
```

then I'd say that yes, that is very helpful, and well worth knowing. However, that being said, there are still some advantages to the way I'm shooting for:

  1. The props checking way won't remain quite as readable in cases where the properties of lots of components need to be checked.
  2. I like the idea that the _then_ clause is consistent between _structure-tests_ (where there is no _when_ clause) and _behaviour-tests_ (where there is).
  3. For large test suites, I can see the ~50ms per-test time that the full render causes becoming problematic, which would likely cause me to switch to using a single _binding-test_ to verify that frequently used parts of the UI are hooked up, allowing all subsequent tests to programmatically achieve the same outcome, at which point I benefit by not having to re-write my _then_ clauses to deal with the fact that we no longer have a full render to check props on.

**Comment by dchambers** [comment]
Hi @jquense,

FYI, I worked around this issue by introducing a third type of test; a _binding-test_. We actually use binding tests in our current knockout test suite, but I was originally hoping to make them optional, and something you only did if performance wasn't good enough. However, as it happens, in the case of React using binding tests helps improve readability too, so this has actually been a net win.

In case you're interested, here's an article I've written that describes the approach I've taken:

  * [Testing React Components with Teaspoon & Unexpected](http://dchambers.github.io/articles/testing-react-components-with-teaspoon-and-unexpected/)
 
Again, thanks for your help figuring all of this out.
